LOOP-1057: Support for in-app and user notification alerts

rickpasetto opened the pull request:
This is a first draft of building out a user alert facility that DeviceManagers can use to post alerts either via UserNotification, via a in-app Modal, or both.  Please note that this is just the first step, there's more work to be done here (on the design side as well), but I think it is better to show this first step so that subsequent steps are smaller PRs.

This is just the Loop part, which includes:
- `UserAlertManager`, the "singleton" which is where we hold onto the `UserAlertHandlers` (see [the `LoopKit` PR](https://github.com/tidepool-org/LoopKit/pull/66)), and eventually have some persistence.  For now it is pretty small.
- Changes to `DeviceDataManager` (basically, this is where I hung `UserAlertManager` -- better to do that than to just dump the functionality in here).  You'll note that it refers to only the interface `UserAlertHandler` that `UserAlertManager` implements -- the goal is to facilitate unit testing, which I'm just getting started on.
- Actual implementations of `UserAlertHandler`: `InAppUserAlertHandler` and `UserNotificationAlertHandler`.  The former is an implementation that simply uses `UIAlertControllers` and a `Timer` to handle triggering.  The latter is pretty much a straight mapping of `UserAlert` to `UserNotification`.

There are a few other minor comments below.  I wanted to make sure there was agreement on the general approach, and then I'd hammer on the unit tests.

rickpasetto committed:
checkpoint

rickpasetto committed:
checkpoint - kind of working

rickpasetto committed:
checkpoint

rickpasetto committed:
checkpoint

rickpasetto committed:
tweak

rickpasetto committed:
unused import

rickpasetto committed:
Fix a bug introduced when I cleaned up the code

nhamming commented on Loop/AppDelegate.swift:
```diff
@@ -122,16 +129,18 @@ extension AppDelegate: UNUserNotificationCenterDelegate {
-                deviceManager.acknowledgeCGMAlert(alertID: alertID)
+        case NotificationManager.Action.acknowledgeAlert.rawValue:
+            let userInfo = response.notification.request.content.userInfo
+            if let alertTypeIdentifier = userInfo[LoopNotificationUserInfoKey.alertID.rawValue] as? UserAlert.TypeIdentifier,
```
[nit] maybe instead of `alertID` use `alertTypeID`?

nhamming commented on Loop/Managers/UserAlert/InAppUserAlertHandler.swift:
```diff
@@ -0,0 +1,113 @@
+            show(alert: alert)
+        case .delayed(let interval):
+            schedule(alert: alert, interval: interval, repeats: false)
+        case .repeating(let interval):
+            schedule(alert: alert, interval: interval, repeats: true)
+        }
```
For a repeating alert, should it be delayed? I mean is there a use case for an alert that shows immediately and also repeats?

rickpasetto replied:
I like that idea, but there isn't a current need yet.  The existing system only uses UserNotifications, and `UNTimeIntervalNotificationTrigger` behaves such that it has an initial delay.  At this time I'm only replacing that system...if there's a requirement for it, it would be a separate PR (and JIRA ticket).

nhamming commented on Loop/Managers/UserAlert/InAppUserAlertHandler.swift:
```diff
@@ -0,0 +1,113 @@
+        }
+    }
+    
+    public func unscheduleAlert(managerIdentifier: String, typeIdentifier: UserAlert.TypeIdentifier) {
+        DispatchQueue.main.async {
+            self.alertsPending.filter {
+                $0.1.identifier == UserAlert.getIdentifier(managerIdentifier: managerIdentifier, typeIdentifier: typeIdentifier)
+            }
+            .forEach { timer, alert in
+                timer.invalidate()
+            }
+        }
+    }
```
Should an unscheduled alert still exist on the `alertsPending` stack?

rickpasetto replied:
No. Good catch. 👍

nhamming commented on Loop/Managers/UserAlert/InAppUserAlertHandler.swift:
```diff
@@ -0,0 +1,113 @@
+        }
+    }
+    
+    public func cancelAlert(managerIdentifier: String, typeIdentifier: UserAlert.TypeIdentifier) {
+        DispatchQueue.main.async {
+            self.alertsShowing.filter {
+                $0.1.identifier == UserAlert.getIdentifier(managerIdentifier: managerIdentifier, typeIdentifier: typeIdentifier)
+            }
+            .forEach { alertController, alert in
+                alertController.dismiss(animated: true)
+            }
+        }
+    }
```
Should a cancelled alert still exist on the `alertsShowing` stack?

rickpasetto replied:
No, good catch again 👍

nhamming commented on Loop/Managers/UserAlert/InAppUserAlertHandler.swift:
```diff
@@ -0,0 +1,113 @@
+            let timer = Timer.scheduledTimer(withTimeInterval: interval, repeats: repeats) { [weak self] timer in
+                self?.show(alert: alert)
+                if !repeats {
+                    self?.alertsPending.removeAll { $0.0 == timer && $0.1.identifier == alert.identifier }
```
[nit] I wonder if this should be part of `unscheduleAlert`?

rickpasetto replied:
"this"...referring to removing from alertsPending?  Yes, it should, and I'll add it. 👍

nhamming replied:
Sorry, I meant if instead of using `removeAll` here call `unscheduleAlert`. Would that be appropriate?

rickpasetto replied:
No it isn't appropriate: same answer to your other similar comment -- I do not want to have the side-effects here of `unscheduleAlert` in the Timer callback.

rickpasetto replied:
No, please see my other comment about side-effets.

nhamming commented on Loop/Managers/UserAlert/InAppUserAlertHandler.swift:
```diff
@@ -0,0 +1,113 @@
+                return
+            }
+            let alertController = self.presentAlert(title: content.title, message: content.body, action: content.acknowledgeAction) {
+                self.alertsShowing.removeAll { $1.identifier == alert.identifier }
```
[nit] is the intent that there is only 1 of each type of alert? If so, maybe a dictionary would help with this?

nhamming replied:
Just looked through the other PR and now I'm not sure what the `identifier` means. See comments there which will obviously feed into comments here.

rickpasetto replied:
The reality is it will be very rare for there to be more than one alert showing.  Rare...but not impossible, hence the array.

`identifier` is a tuple of `( managerIdentifier, typeIdentifier)`, in order to uniquely identify an alert.  You're right, I put comments in `UserAlert.swift` for `managerIdentifier` and `typeIdentifier` but not `identifier`.  Its purpose is for "de-duping" alerts when shown. I'll make that clearer, thanks!

nhamming replied:
Follow-up question, can the same type of alert exist in the array multiple times?

rickpasetto replied:
No, the same alert should not exist in the array multiple times.

nhamming replied:
then why not use a dictionary to ensure this doesn't happen and also make the searching more efficient?

rickpasetto replied:
It isn't necessary.  And, though I'd claim it would be a case of premature optimization, it still turns out dictionary lookup would actually be _slower_.  For arrays of less than about 7-10, iterating through an array is faster than a hash value evaluation.

nhamming replied:
It will definitely help ensure there is only 1 alert type in the stack. Also, isn't a dictionary search a O(1) operation while an unordered array search is O(n)?

rickpasetto replied:
You are correct as to the functional order (O-notation), but I encourage you to try it yourself: Try timing a dictionary lookup of a reasonably-sized string hash value calculation vs. an array of strings less than 7.
I'll change it by your request, but I still think I'm right :)

nhamming reviewed (commented):
Generally looks good. Just questions/comment to help me better understand.

rickpasetto committed:
PR Feedback

rickpasetto committed:
PR Feedback

rickpasetto committed:
Wrap up the `managerIdentifier` and `typeIdentifier` into a single type, `Identifier`

rickpasetto commented:
@ps2  @nhamming I've updated the PR with some updates, including introducing `UserAlert.Identifier` like I mention above.  Please take a look.
Unless you object to the general approach here, I will be starting on unit tests.

rickpasetto committed:
PR Feedback: renaming:

UserAlert* -> DeviceAlert*
scheduleAlert -> issueAlert
unscheduleAlert -> removePendingAlerts
cancelAlert -> removeDeliveredAlerts

rickpasetto committed:
More renaming, move acknowledgeDeviceAlert to DeviceAlertManager (where it really should've been)

rickpasetto committed:
Project file, removeAlertResponder

rickpasetto commented on Loop/Managers/DeviceAlert/DeviceAlertManager.swift:
```diff
@@ -0,0 +1,54 @@
+    
+    public func acknowledgeDeviceAlert(managerIdentifier: String, alertTypeIdentifier: DeviceAlert.TypeIdentifier) {
+        if let responder = responders[managerIdentifier]?.value {
+            responder.acknowledgeAlert(typeIdentifier: alertTypeIdentifier)
```
Note: it isn't logging the device event here (anymore), but it is probably redundant anyway.

rickpasetto committed:
More PR Feedback: get rid of AcknowledgeCompletion

rickpasetto committed:
added a comment

nhamming commented on Loop/AppDelegate.swift:
```diff
@@ -138,9 +139,9 @@ extension AppDelegate: UNUserNotificationCenterDelegate {
+            if let alertTypeIdentifier = userInfo[LoopNotificationUserInfoKey.alertTypeId.rawValue] as? DeviceAlert.TypeIdentifier,
                 let managerIdentifier = userInfo[LoopNotificationUserInfoKey.managerIDForAlert.rawValue] as? String {
-                deviceDataManager.acknowledgeDeviceAlert(managerIdentifier: managerIdentifier, alertTypeIdentifier: alertTypeIdentifier)
+                deviceAlertManager.acknowledgeDeviceAlert(deviceManagerInstanceIdentifier: managerIdentifier, alertTypeIdentifier: alertTypeIdentifier)
```
Does this really identify an instance of a device manager? I would have to look again, but I understood that it identifies a type of manager, but is not sufficient to distinguish between say 2 dash managers.

rickpasetto replied:
Now that I think about it (and after sleeping on it), I think I'm agreeing with you that I don't like the `Instance` part of this name.  I will rename it everywhere by removing it.   I hope @ps2 still likes it...

nhamming commented on Loop/Managers/DeviceAlert/InAppModalDeviceAlertHandler.swift:
```diff
@@ -72,17 +76,18 @@ extension InAppUserAlertHandler {
-            let alertController = self.presentAlert(title: content.title, message: content.body, action: content.acknowledgeAction) {
+            let alertController = self.presentAlert(title: content.title, message: content.body, action: content.acknowledgeActionButtonLabel) {
                 self.alertsShowing.removeAll { $1.identifier == alert.identifier }
-                alert.acknowledgeCompletion?(alert.typeIdentifier)
```
[nit] could you replace `removeAll` with `removeDeliveredAlerts`?

rickpasetto replied:
Not really, because I don't want the other side-effects from `removeDeliveredAlerts` (in this case, I don't need to call `dismiss` in the dismiss-completion callback here...)

nhamming commented on Loop/Managers/DeviceAlert/InAppModalDeviceAlertHandler.swift:
```diff
@@ -31,33 +33,35 @@ public class InAppUserAlertHandler: UserAlertHandler {
     }
     
-    public func unscheduleAlert(managerIdentifier: String, typeIdentifier: UserAlert.TypeIdentifier) {
+    public func removePendingAlerts(identifier: DeviceAlert.Identifier) {
```
[nit] By adding the `s` this helps me understand that there may be the same type of alert for a specific manager in the stack multiple times. Is this the intention?

rickpasetto replied:
This was renamed per @ps2 's suggestion.  But, I think I agree with you: it should not be possible to put up more than one alert with the same identifier.  I'll remove the 's'.

nhamming commented on Loop/Managers/DeviceAlert/InAppModalDeviceAlertHandler.swift:
```diff
@@ -31,33 +33,35 @@ public class InAppUserAlertHandler: UserAlertHandler {
     }
     
-    public func cancelAlert(managerIdentifier: String, typeIdentifier: UserAlert.TypeIdentifier) {
+    public func removeDeliveredAlerts(identifier: DeviceAlert.Identifier) {
```
[nit] By adding the `s` this helps me understand that there may be the same type of alert for a specific manager in the stack multiple times. Is this the intention?

rickpasetto replied:
see above

nhamming reviewed (commented):
I'm still a bit confused about the alert stacks and what could be found in them. But don't let that stop you. This is mostly nits with a question that was also asked with additional suggestions in `LoopKit`.

rickpasetto committed:
PR feedback: renaming only

deviceManagerInstanceIdentifier -> deviceManagerIdentifier
removePendingAlerts -> removePendingAlert
removeDeliveredAlerts -> removeDeliveredAlert

rickpasetto committed:
More PR Feedback:

- Made `alertsShowing` and `alertsPending` a dictionary, by request.  Also took the opportunity to add some bulletproofing "preconditions".
- Also, a bonus: I didn't like the asymmetry of the `DeviceAlertManagerResponder` protocol, and it was unnecessary.  So a small change there.

rickpasetto committed:
PR feedback: one more rename: deviceManagerIdentifier -> managerIdentifier

nhamming reviewed (approved):
LGTM

ps2 commented on Loop/AppDelegate.swift:
```diff
@@ -19,31 +19,36 @@ final class AppDelegate: UIResponder, UIApplicationDelegate {
 
     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
-        SharedLogging.instance = deviceManager.loggingServicesManager
+        deviceAlertManager = DeviceAlertManager(rootViewController: rootViewController,
```
This will need to be revisited when we bring the changes to deal with potential BFU launching from DIY in, as the root vc construction is delayed, iirc. But we can deal with that then, I suppose.

rickpasetto replied:
Yeah, I'd have to check, but I _think_ we can just defer this code too

ps2 commented on Loop/Managers/NotificationManager.swift:
```diff
@@ -14,7 +14,7 @@ struct NotificationManager {
     enum Action: String {
         case retryBolus
-        case acknowledgeCGMAlert
+        case acknowledgeAlert
```
acknowledgeDeviceAlert?

ps2 commented on Loop/Managers/DeviceAlert/DeviceAlertManager.swift:
```diff
@@ -0,0 +1,62 @@
+/// - etc.
+public final class DeviceAlertManager {
+
+    var handlers: [DeviceAlertHandler] = []
```
Looking at it in this context, perhaps "DeviceAlertPresenter" is more specific to the role being performed.

ps2 commented on Loop/Managers/DeviceAlert/InAppModalDeviceAlertHandler.swift:
```diff
@@ -0,0 +1,141 @@
+    }
+    
+    // Helper function pulled from SO...may be outdated, especially in the SwiftUI world
+    private func topViewController(controller: UIViewController?) -> UIViewController? {
```
Did you verify that this was needed? I.e. that presenting from the rootViewController directly results in an obscured alert?

rickpasetto replied:
I'm not sure in every case in our app, and I certainly am not sure about every SwiftUI screen. This was a suggestion from StackOverflow.

ps2 commented on Loop/Managers/DeviceAlert/UserNotificationDeviceAlertHandler.swift:
```diff
@@ -0,0 +1,85 @@
+        let userNotificationContent = UNMutableNotificationContent()
+        userNotificationContent.title = content.title
+        userNotificationContent.body = content.body
+        userNotificationContent.sound = content.isCritical ? .defaultCritical : .default
```
We do need to deal with custom sounds per the Dexcom integration. We don't have the actual sound files yet, afaik, but this facility will need to be revamped to support custom sounds. That might be out of the scope of this ticket. If so, do we have a ticket covering this? We know we need it.

rickpasetto replied:
https://tidepool.atlassian.net/browse/LOOP-1179

ps2 commented on Loop/Managers/NotificationManager.swift:
```diff
@@ -33,17 +33,17 @@ struct NotificationManager {
-            identifier: Action.acknowledgeCGMAlert.rawValue,
-            title: NSLocalizedString("OK", comment: "The title of the notification action to acknowledge a cgm alert"),
-            options: []
+        let acknowledgeAlertAction = UNNotificationAction(
```
acknowledgeDeviceAlertAction?
